src/cmsisPosix_Semaphore.c

```diff
@@ -0,0 +1,128 @@
+#define _GNU_SOURCE
+#include <semaphore.h>
+#include <pthread.h>
+#include <errno.h>
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include "cmsis_os2.h"
+#include "cmsisPosix_Config.h"
+
+// Forward declarations
+void cp_timeoutToTimespec(uint32_t timeout, struct timespec *ts);
+
+// Structure to hold POSIX semaphore and CMSIS attributes
+typedef struct
+{
+    const char *name;   // Name of semaphore
+    sem_t sem;          // POSIX semaphore
+} cp_semaphoreData_t;
+
+osSemaphoreId_t osSemaphoreNew(uint32_t max_count, uint32_t initial_count, const osSemaphoreAttr_t *attr)
+{
+    if (max_count == 0) {
+        return NULL;
+    }
+
+    cp_semaphoreData_t *semaphore = malloc(sizeof(cp_semaphoreData_t));
+    if (semaphore == NULL) {
+        return NULL;
+    }
+
+    if (sem_init(&semaphore->sem, 0, max_count) != 0) {
+        free(semaphore);
+        return NULL;
+    }
+
+    semaphore->name = (attr ? attr->name : NULL);
+
+    // Lock semaphore until its value is less than or equals initial_count
+    while (osSemaphoreGetCount(semaphore) > initial_count) {
```

Comment:
As written in other comment, sem_init getting initial value hence there is no need to this operation here